From fa096dd7e7a23e766c181b4878a537fe94f3c415 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Sun, 1 Sep 2019 03:22:12 +0200
Subject: [PATCH 10/26] thermal: sun8i: Lower rate of sensor data interrupts to
 4/sec/sensor

4 readings per second are more than enough. Previously there were on
the order of 30-50 per second.

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 drivers/thermal/sun8i_thermal.c | 34 +++++++++++++++++++++++----------
 1 file changed, 24 insertions(+), 10 deletions(-)

diff --git a/drivers/thermal/sun8i_thermal.c b/drivers/thermal/sun8i_thermal.c
index e510c979fa39..34b599afb7bb 100644
--- a/drivers/thermal/sun8i_thermal.c
+++ b/drivers/thermal/sun8i_thermal.c
@@ -391,17 +391,22 @@ static int sun8i_h3_thermal_init(struct ths_device *tmdev)
 		     SUN50I_THS_FILTER_EN |
 		     SUN50I_THS_FILTER_TYPE(1));
 	/*
-	 * period = (x + 1) * 4096 / clkin; ~10ms
-	 * enable data interrupt
+	 * clkin = 24MHz
+	 * filter_samples = 4
+	 * period = 0.25s
+	 *
+	 * x = period * clkin / 4096 / filter_samples - 1
+	 *   = 365
 	 */
 	val = GENMASK(7 + tmdev->chip->sensor_num, 8);
 	regmap_write(tmdev->regmap, SUN8I_THS_IC,
-		     SUN50I_H6_THS_PC_TEMP_PERIOD(58) | val);
+		     SUN50I_H6_THS_PC_TEMP_PERIOD(365) | val);
 	/*
+	 * T_acq = 20us
 	 * clkin = 24MHz
-	 * T acquire = clkin / (x + 1)
-	 *           = 20us
-	 * enable sensor
+	 *
+	 * x = T_acq * clkin - 1
+	 *   = 479
 	 */
 	regmap_write(tmdev->regmap, SUN8I_THS_CTRL0,
 		     SUN8I_THS_CTRL0_T_ACQ0(479));
@@ -417,9 +422,11 @@ static int sun50i_h6_thermal_init(struct ths_device *tmdev)
 	int val;
 
 	/*
+	 * T_acq = 20us
 	 * clkin = 24MHz
-	 * T acquire = clkin / (x + 1)
-	 *           = 20us
+	 *
+	 * x = T_acq * clkin - 1
+	 *   = 479
 	 */
 	regmap_write(tmdev->regmap, SUN50I_THS_CTRL0,
 		     SUN50I_THS_CTRL0_T_ACQ(479));
@@ -427,9 +434,16 @@ static int sun50i_h6_thermal_init(struct ths_device *tmdev)
 	regmap_write(tmdev->regmap, SUN50I_H6_THS_MFC,
 		     SUN50I_THS_FILTER_EN |
 		     SUN50I_THS_FILTER_TYPE(1));
-	/* period = (x + 1) * 4096 / clkin; ~10ms */
+	/*
+	 * clkin = 24MHz
+	 * filter_samples = 4
+	 * period = 0.25s
+	 *
+	 * x = period * clkin / 4096 / filter_samples - 1
+	 *   = 365
+	 */
 	regmap_write(tmdev->regmap, SUN50I_H6_THS_PC,
-		     SUN50I_H6_THS_PC_TEMP_PERIOD(58));
+		     SUN50I_H6_THS_PC_TEMP_PERIOD(365));
 	/* enable sensor */
 	val = GENMASK(tmdev->chip->sensor_num - 1, 0);
 	regmap_write(tmdev->regmap, SUN50I_H6_THS_ENABLE, val);
-- 
2.23.0

